### ISSUE TITLE
[Q] Using the API's in multiple threads

### ISSUE THREAD

**Issue by pfedor** [googleapis/google-cloud-cpp#5668]
Hello,

pubsub::Publisher, bigtable::Table and spanner::Client all have comments like this:

"Instances of this class created via copy-construction or copy-assignment share the underlying pool of connections. Access to these copies via multiple threads is guaranteed to work. Two threads operating on the same instance of this class is not guaranteed to work."

I want to make sure I understand what it means. What it seems to say is that

(a) pubsub::Publisher etc. are not thread-safe, and

(b) one should make a copy of a Publisher before using it in a multi-threaded situation (e.g., in an rpc handler). (As opposed to, say, guarding it with a mutex).

In other words, if I want to use a Publisher in an rpc handler, using grpc, it seems to me the documentation suggests the following pattern:

```
class GeometryServiceImpl final : public Geometry::Service {
 public:
  GeometryServiceImpl(std::shared_ptr<pubsub::PublisherConnection> pubsub_conn)
      : publisher_(pubsub_conn) {}

  grpc::Status ScheduleLengthComputation(
      grpc::ServerContext* context,
      const ScheduleLengthComputationRequest* request,
      ScheduleLengthComputationResponse* response) override {
    auto publisher = publisher_;
    auto message_id = publisher
                          .Publish(pubsub::MessageBuilder()
                                       .SetData(request->SerializeAsString())
                                       .Build())
                          .get();
    if (!message_id.ok()) {
      return grpc::Status(
          static_cast<grpc::StatusCode>(message_id.status().code()),
          message_id.status().message() +
              "; publishing a length computation request to pubsub.");
    }
    return grpc::Status::OK;
  }

 private:
  const pubsub::Publisher publisher_;
};
```
and
```
  std::shared_ptr<pubsub::PublisherConnection> pubsub_conn(
      pubsub::MakePublisherConnection(pubsub::Topic(kProjectId, kTopicId), {}));

  GeometryServiceImpl service(pubsub_conn);

  //... start the server
```

Is the above correct? Is that the pattern I should be using for pubsub::Publisher, bigtable::Table and spanner::Client?

Thanks!

**Comment by coryan** [comment]
> (a) pubsub::Publisher etc. are not thread-safe, and

This is correct.  That is because there are operations (e.g. assignment) on `pubsub::Publisher` that change the state without any synchronization.

> (b) one should make a copy of a Publisher before using it in a multi-threaded situation (e.g., in an rpc handler). (As opposed to, say, guarding it with a mutex).

**TL;DR;** I think your code is safe today, and I cannot think of any changes that we would make to break your code.

Your code works because `publisher_` is declared as `const pubsub::Publisher`, so it cannot be changed once created *and* you copy it before using it.  But if `publisher_` was not `const`, you would need to protect it with a mutex or some other synchronization before making a copy of it.

In general, what we want to warn folks about this:

```cc

void worker(pubsub::Publisher& p1, pubsub::Publisher p2) {
  auto m1 = p1.Publish(...); // ERROR, publisher is shared by multiple threads
  auto m2 = p2.Publish(...); // OK, publisher not shared
}

void parallel(pubsub::Publisher p) {
  auto t1 = std::thread(worker, std::ref(p), p);
  auto t2 = std::thread(worker, std::ref(p), p);
  t1.join();
  t2.join();
}
```

**Comment by pfedor** [comment]
Thank you for the answer! Just to confirm, same applies to ```bigtable::Table``` and ```spanner::Client```, correct?

**Comment by coryan** [comment]
> Thank you for the answer! Just to confirm, same applies to `bigtable::Table` and `spanner::Client`, correct?

Yes.

**Comment by pfedor** [comment]
Thank you very much.
